test/faq.cpp

```diff
@@ -0,0 +1,184 @@
+  for (int z = 0; z < depth; z++) {
+    for (int y = 0; y < height; y++) {
+      for (int x = 0; x < width; x++) {
+        c_array[z][y][x] = std::make_tuple(x, y, z);
+      }
+    }
+  }
+
+  // Now we can create an array_ref of this c_array, and check that
+  // the coordinates map to the same location, indicating that we used
+  // the same convention to determine strides as the C compiler:
+  dense_array_ref<std::tuple<int, int, int>, 3> c_array_ref(&c_array[0][0][0], {width, height, depth});
+  for_each_index(c_array_ref.shape(), [&](std::tuple<int, int, int> i) {
+    ASSERT_EQ(c_array_ref[i], i);
+  });
+}
+
+TEST(faq_reshape) {
+  // Q: How do I resize or change the shape of an already constructed array?
+
+  // A: There are several options, with different behaviors. First, we can use
+  // `array::reshape`, which changes the shape of an array while moving the
+  // elements of the intersection of the old shape and new shape to the new
+  // array, similar to `std::vector::resize`:
+  dense_array<int, 2> a({3, 4});
+  for_all_indices(a.shape(), [&](int x, int y) {
+    a(x, y) = y * 3 + x;
+  });
+  for (auto y : a.y()) {
+    for (auto x : a.x()) {
+      std::cout << a(x, y) << " ";
+    }
+    std::cout << std::endl;
+  }
+  std::cout << std::endl;
+  // Output:
+  //   0 1 2
+  //   3 4 5
+  //   6 7 8
+  //   9 10 11
+
+  a.reshape({2, 6});
+  for (auto y : a.y()) {
+    for (auto x : a.x()) {
+      std::cout << a(x, y) << " ";
+    }
+    std::cout << std::endl;
+  }
+  std::cout << std::endl;
+  // Output:
+  //   0 1
+  //   3 4
+  //   6 7
+  //   9 10
+  //   0 0
+  //   0 0
+  // Observe that the right column of the original array has been lost, and
+  // two default-constructed rows have been added to the bottom of the array.
+
+  // A: We can also reinterpret the shape of an existing array using
+  // `array::set_shape`:
+  a.set_shape({4, 3});
+  for (auto y : a.y()) {
+    for (auto x : a.x()) {
+      std::cout << a(x, y) << " ";
+    }
+    std::cout << std::endl;
+  }
+  std::cout << std::endl;
+  // Output:
+  //   0 1 3 4
+  //   6 7 9 10
+  //   0 0 0 0
+  // Observe that this has not removed or added any values from the array, the
+  // underlying memory has simply be reinterpreted as a different array.
+
+  // A: We can also use `reinterpret_shape` to make a new `array_ref` with the
+  // new shape:
+  auto a_reshaped = reinterpret_shape(a, dense_shape<2>{4, 3});
+  ASSERT(a_reshaped == a);
+
+  // Q: `array`'s move constructor requires the source array to have the same
+  // shape type. How do I move ownership to an array of a different shape type?
+
+  // A: The helper function `move_reinterpret_shape` combines move construction
+  // with `reinterpret_shape`:
+  array_of_rank<int, 3> source({3, 4, 5});
+  dense_array<int, 3> dest = move_reinterpret_shape<dense_shape<3>>(std::move(source));
+
+  // This can fail at runtime if the source shape is not compatible with the
+  // destination shape.
+}
+
```

Comment:
How about a few more FAQs on what shape::resolve() does, and what shapes or array_refs can be passed to functions with automatic conversion?

Reply:
Hmm, can you share a case in which you actually needed to care about resolve? I think it's not something that normally you should have to think about.